streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -634,29 +634,32 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
       for {
         input       <- SingleProducerAsyncInput.make[InErr, InElem, InDone]
         queueReader  = ZChannel.fromInput(input)
-        queue       <- Queue.bounded[ZIO[Env1, OutErr1, Either[OutDone, OutElem2]]](n)
-        _           <- scope.addFinalizer(queue.shutdown)
-        errorSignal <- Promise.make[OutErr1, Nothing]
+        outgoing    <- Queue.bounded[ZIO[Env1, Unit, Either[OutDone, OutElem2]]](n)
+        _           <- scope.addFinalizer(outgoing.shutdown)
+        errorSignal <- Promise.make[Unit, Nothing]
         permits     <- Semaphore.make(n.toLong)
+        failure     <- Ref.make[Cause[OutErr1]](Cause.empty)
         pull        <- (queueReader >>> self).toPullIn(scope)
         _ <- pull
                .foldCauseZIO(
-                 cause => queue.offer(ZIO.refailCause(cause)),
+                 cause => failure.update(_ && cause) *> outgoing.offer(ZIO.fail(())) *> ZIO.fail(()),
```

Comment:
Same as above, consider using `refailCause(Cause.unit)` which is much more lightweight as it doesn't generate fiber stack traces

Reply:
Nice, I didn't know about that